test/cpp/end2end/xds_end2end_test.cc

```diff
@@ -1652,7 +1835,71 @@ TEST_F(SingleBalancerWithClientLoadReportingTest, BalancerRestart) {
   EXPECT_EQ(0U, client_stats->total_dropped_requests());
 }
 
-// TODO(juanlishen): Add TEST_F(SingleBalancerWithClientLoadReportingTest, Drop)
+TEST_F(SingleBalancerWithClientLoadReportingTest, Drop) {
+  SetNextResolution({}, kDefaultServiceConfig_.c_str());
+  SetNextResolutionForLbChannelAllBalancers();
+  const size_t kNumRpcs = 1000;
+  const uint32_t drop_per_million_for_lb = 100000;
+  const uint32_t drop_per_million_for_throttle = 200000;
+  const double drop_rate_for_lb = drop_per_million_for_lb / 1000000.0;
+  const double drop_rate_for_throttle =
+      drop_per_million_for_throttle / 1000000.0;
+  const double drop_rate_for_lb_and_throttle =
+      drop_rate_for_lb + (1 - drop_rate_for_lb) * drop_rate_for_throttle;
+  // The EDS response contains two drop categories.
+  ScheduleResponseForBalancer(
+      0,
+      EdsServiceImpl::BuildResponse(
+          GetBackendPortsInGroups(),
+          {{"lb", drop_per_million_for_lb},
+           {"throttle", drop_per_million_for_throttle}}),
+      0);
+  int num_ok = 0;
+  int num_failure = 0;
+  int num_drops = 0;
+  std::tie(num_ok, num_failure, num_drops) = WaitForAllBackends();
+  const size_t num_warmup = num_ok + num_failure + num_drops;
+  // Send kNumRpcs RPCs and count the drops.
+  for (size_t i = 0; i < kNumRpcs; ++i) {
+    EchoResponse response;
+    const Status status = SendRpc(&response);
+    if (!status.ok() &&
+        status.error_message() == "Call dropped by load balancing policy") {
+      ++num_drops;
+    } else {
+      EXPECT_TRUE(status.ok()) << "code=" << status.error_code()
+                               << " message=" << status.error_message();
+      EXPECT_EQ(response.message(), kRequestMessage_);
+    }
+  }
+  // The drop rate should be roughly equal to the expectation.
+  const double seen_drop_rate = static_cast<double>(num_drops) / kNumRpcs;
+  const double error_tolerance = 0.2;
+  EXPECT_THAT(seen_drop_rate,
+              ::testing::AllOf(::testing::Ge(drop_rate_for_lb_and_throttle *
+                                             (1 - error_tolerance)),
+                               ::testing::Le(drop_rate_for_lb_and_throttle *
+                                             (1 + error_tolerance))));
+  // Check client stats.
+  ClientStats* client_stats = balancers_[0]->lrs_service()->WaitForLoadReport();
+  EXPECT_EQ(num_drops, client_stats->total_dropped_requests());
+  const size_t total_rpc = num_warmup + kNumRpcs;
+  EXPECT_THAT(
+      client_stats->dropped_requests("lb"),
+      ::testing::AllOf(
+          ::testing::Ge(total_rpc * drop_rate_for_lb * (1 - error_tolerance)),
+          ::testing::Le(total_rpc * drop_rate_for_lb * (1 + error_tolerance))));
+  EXPECT_THAT(
+      client_stats->dropped_requests("throttle"),
+      ::testing::AllOf(
+          ::testing::Ge(total_rpc * (1 - drop_rate_for_lb) *
```

Comment:
I don't think we should need to account for `drop_rate_for_lb` when looking at `drop_rate_for_throttle`.  See my comment elsewhere about not wanting to do a different roll of the dice for each drop category.